### PR TITLE
Support gradient masking in acoustic wave propagation

### PR DESCRIPTION
**Pull request by BlissChapman**
# Motivation
For some inversion problems, a subset of the true density/velocity model is known. In these use cases, we only need to solve for the subset of the model parameters that are unknown. Large memory savings are possible from only storing the gradient where necessary.

This pull request introduces a gradient mask parameter to the acoustic wave propagator that allows the CUDA kernel to compute gradients only for a subset of grid points using a sparse gradient mask index:
```
        gradient_mask: A boolean Tensor with the same spatial shape as the
            model, specifying which cells should have gradients computed.
            Optional. If not provided, gradients will be computed everywhere in
            the model. If the model is padded internally, an unpadded mask will
            be padded with False values. True values indicate cells where
            gradients should be computed, while False values indicate cells
            where gradients should be set to 0.
```

# Test Plan
Added new unit tests:
1. Test that gradient masking reduces memory requirements.
2. Test that the gradient mask zeros gradients outside the mask.
3. Test that providing no gradient mask computes the gradient everywhere.
4. Test that gradient masking is not supported when using the python backend.

```
pip install . && PYTHONPATH= python -m pytest
```

In addition, a similar version of this code has been used internally at my work for full waveform inversion with a computational domain that cannot fit in H200 gpu memory without gradient masking.

### CODE REVIEW

**Comment by ar4** [comment]
Thank you, Bliss. This is great. It's an elegant implementation, and you did a great job of keeping the style consistent. A few comments:

### 1. Testing

I suggest making your tests a bit more sensitive to possible problems. For example, the following code:
```python
from typing import Optional, Tuple

import torch

from deepwave import acoustic


def _compute_gradient(
    mask: Optional[torch.Tensor], nx: Tuple[int, int] = (5, 5), nt: int = 4
) -> Tuple[torch.Tensor, torch.Tensor]:
    v = torch.ones(nx, requires_grad=True)
    rho = torch.ones_like(v, requires_grad=True)
    source_amplitudes = torch.zeros(1, 1, nt)
    source_amplitudes[0, 0, 0] = 1
    source_locations = torch.tensor([[[0, 0]]], dtype=torch.long)
    receiver_locations = torch.tensor([[[nx[0] - 1, nx[1] - 1]]], dtype=torch.long)
    outputs = acoustic(
        v,
        rho,
        1.0,
        0.001,
        source_amplitudes_p=source_amplitudes,
        source_locations_p=source_locations,
        receiver_locations_p=receiver_locations,
        nt=nt,
        pml_width=2,
        gradient_mask=mask,
    )
    receivers = outputs[-3:]
    loss = sum(o.sum() for o in receivers)
    loss.backward()
    assert v.grad is not None
    assert rho.grad is not None
    return v.grad.detach(), rho.grad.detach()


nx = (10, 10)
mask = torch.zeros(nx, dtype=torch.bool)
mask.fill_(True)
# mask[1:-1, 1:-1] = True
grad_no_mask_v, grad_no_mask_rho = _compute_gradient(None, nx=nx, nt=500)
grad_masked_v, grad_masked_rho = _compute_gradient(mask, nx=nx, nt=500)
grad_no_mask_v[~mask] = 0
grad_no_mask_rho[~mask] = 0
```
reveals that the gradients at the edges of the model do not exactly match the unmasked gradient.
mask_all
I think this can be resolved by changing [the padding of the mask](https://github.com/ar4/deepwave/blob/7543997a484ffd02319dd241e4e3befbbb541833/src/deepwave/acoustic.py#L507) to be replication rather than zero-padding, so that the gradients will be calculated in the PML regions when the mask touches the edge of the model.

I also suggest testing the case when the mask does not cover the whole model. Using `mask[1:-1, 1:-1] = True` instead of `mask.fill_(True)` shows that the density gradients at the edges of the mask region do not match exactly.
mask_part
I think this could be resolved by extending the gradient mask by one cell in each dimension (where possible), so that all of the buoyancy gradient components that contribute to the density gradient are calculated.

### 2. Vectorization

The compiler is having greater difficulty vectorizing the C code because of the conditionals in the inner loop. On my system, this causes the CPU runtime of the propagator to double, even when the gradient masking feature is not used. One possible way to overcome this would be to use the normal full gradients within the inner loop, but then change the storing and loading code so that the masking occurs there, outside the main spatial computation loops. That would, however, interfere with vectorization of the loading/storing code, so I'm not sure how much of an improvement it would be.

### 3. Consider improving existing alternatives (callbacks, compression)

Two alternatives to masked gradients in the C/CUDA code are:
  (a) using callback functions to achieve the same effect
  (b) storing the entire gradient but using the new storage features to ease the memory constraints

In my tests comparing the runtime (I only tested 2D), the runtime of (a) was about 1.5-2 times that of the C/CUDA masked gradients, and the runtime of (b) was only slightly higher on the CPU but was about four times that of the CUDA masked gradients code on the GPU. These approaches both have the advantage of not adding complexity to the existing code or any performance penalty to those who do not use masked gradients. It therefore might be worth considering whether improving the performance of either of these options is possible as an alternative to adding the masked gradients feature.

At the moment control returns to the Python code to call the callbacks. Although this is done within an autograd.Function, avoiding the overhead of PyTorch's automatic gradient calculation, there is still some overhead. I suspect that this overhead is the main reason why the callback version is slower than the C/CUDA masked gradient implementation. I will do some tests over the next few days to try to measure this overhead. If this overhead is confirmed to be the source of the slower runtime, there are some possible solutions. One is to pass a function pointer to the callback function into the C/CUDA code (using ctypes), enabling the C/CUDA code to directly call the callback function without having to go back to the propagator's Python wrapper. The other is to reduce the overhead of calling the C/CUDA code from Python. Deepwave currently uses ctypes to do this (because of its portability), but there are other means of calling C/CUDA from Python that apparently have lower overhead.

I added double buffering and CUDA streams to Deepwave a few days ago to improve the performance of some of the new storage options (compression and storage to CPU memory). Further profiling might be useful to identify why it is still so slow. The current compression method is quite simplistic, so I suspect that improvements are possible.  It does reduce the memory requirement by about three quarters, however, and storage in CPU memory further reduces the constraint imposed by limited GPU memory, so in some cases this might be a sufficient alternative to masked gradients.

### 4. Consider different method (local solver)

Another option to consider, although it is a larger undertaking, is to investigate something like [this local solver approach](https://scholar.google.com/citations?view_op=view_citation&hl=en&user=pkmHvCQAAAAJ&citation_for_view=pkmHvCQAAAAJ:eQOLeE2rZwMC), which might also reduce runtime.

### Summary

Thank you again for all of your work on this. To summarise, the main issues to be resolved at the moment are the vectorization of the C code and the differences between the masked gradient and the regular gradient in the masked areas. I will investigate whether the runtime of the two existing alternatives (callbacks and compressed storage) can be improved, and I suggest you consider whether a local solver-type approach might be a good fit for your application.
